### PR TITLE
Fix einsum gradient by using old TF einsum

### PR DESCRIPTION
**Pull request by trbromley**
TF2.1+ introduces a bug into `einsum` that results in an incorrect gradient for complex numbers: https://github.com/tensorflow/tensorflow/issues/37307.

This PR adds a quick fix where the old version of `einsum` will be used for TF2.1+.

### CODE REVIEW

File: strawberryfields/backends/tfbackend/circuit.py [hunk 1]
```diff
@@ -38,6 +38,13 @@
 from scipy.special import factorial
 import tensorflow as tf
 
+try:
+    from tensorflow.python.ops.special_math_ops import _einsum_v1
+
+    tf.einsum = _einsum_v1
+except ImportError:
+    pass
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/circuit.py:46]: With TF 2.1, the legacy `tf.einsum` was renamed to `_einsum_v1`, while the replacement `tf.einsum` introduced the bug. This `try-except` block will dynamically patch versions where `_einsum_v1` exists, to make it the default einsum implementation _within this file only_.

File: tests/integration/test_tf_integration.py [hunk 2]
```diff
@@ -422,3 +422,25 @@ def test_coherent_dm_gradient(self, setup_eng, cutoff, tol, batch_size):
         grad = tape.gradient(res, [a])
         grad_ex = -a ** 3 * (a ** 2 - 2) * np.exp(-a ** 2)
         assert np.allclose(grad, grad_ex, atol=tol, rtol=0)
+
+
+def test_TODO():
```
Comment (co9olguy) [inline tests/integration/test_tf_integration.py:427]: I'm curious about this test name :thinking:

File: tests/integration/test_tf_integration.py [hunk 3]
```diff
@@ -422,3 +422,25 @@ def test_coherent_dm_gradient(self, setup_eng, cutoff, tol, batch_size):
...
+        return tf.abs(tf.reduce_sum(_einsum_v1('ab,bc->ac', h, h)))
+
+    A = tf.Variable(tf.random.uniform(shape=[2, 2]))
+    B = tf.random.uniform(shape=[2, 2])
+    for f in (f1, f2, f3):
+        with tf.GradientTape() as tape:
+            C = tf.cast(A, dtype=tf.complex64) + 1j * tf.cast(B, dtype=tf.complex64)
+            loss = f(C)
+        grad = tape.gradient(loss, A)
+        print(grad)
```
Comment (co9olguy) [inline tests/integration/test_tf_integration.py:446]: shouldn't a test conclude with an `assert`, not a `print`?

File: tests/integration/test_tf_integration.py (hunk 2 above)
Comment (co9olguy) [reply tests/integration/test_tf_integration.py:427]: No docstring either

**Comment by co9olguy** [comment]
I'm glad to have this hotfix in, but can we give the commensurate test a bit more attention?

**Comment by josh146** [comment]
@co9olguy, the test was updated and available here: https://github.com/XanaduAI/strawberryfields/blob/c7dd6f58fb6b8a9cb306de4b2e4e9ef13824d2a1/tests/integration/test_tf_integration.py#L484

File: tests/integration/test_tf_integration.py (hunk 2 above)
Comment (trbromley) [reply tests/integration/test_tf_integration.py:427]: Yes this shouldn't be here (yet), I think it was a bit of a miscommunication on the PR being ready to merge.

The idea of this test is to compare gradients using old and new versions of einsum. The plan is to assert that the gradients should be the same and add an xfail to show that we expect failure - when this stops being the case we know they've fixed the bug.

Unfortunately I couldn't get it to work and was going to come back to this today, hence the TODO and the print statements - it's just in an early prototype form.

**Comment by co9olguy** [comment]
Thanks

File: tests/integration/test_tf_integration.py (hunk 2 above)
Comment (trbromley) [reply tests/integration/test_tf_integration.py:427]: Since this is merged into the `tf2` branch and has not yet made it to `master`, we can sort this out by doing a follow up PR to properly add the test.

Comment (trbromley) [reply tests/integration/test_tf_integration.py:427]: Oh wait, @josh146 already fixed it! https://github.com/XanaduAI/strawberryfields/commit/932856b9e399331be7966c7d962964f8d750f361 :100:
